File: assistant/src/home/relationship-state.ts
/**
 * Relationship state data contract.
 *
 * This is the shared wire format that describes the assistant's current
 * relationship with the user: which tier they're at, what facts the
 * assistant has learned about them, and which capabilities are unlocked.
 *
 * The TypeScript types here are the source of truth. A structurally
 * identical Swift definition lives at
 * `clients/shared/Models/RelationshipState.swift` — any change here must
 * be mirrored there (and the contract test guards the default list).
 */

export const RELATIONSHIP_STATE_VERSION = 1 as const;

export type RelationshipTier = 1 | 2 | 3 | 4;

export interface TierInfo {
  label: string;
  description: string;
  nextTierHint?: string;
}

export const TIER_INFO: Record<RelationshipTier, TierInfo> = {
  1: {
    label: "Getting to know you",
    description: "We just met — learning the basics",
    nextTierHint: "A few more conversations and I'll start to find my footing",
  },
  2: {
    label: "Finding my footing",
    description: "Starting to understand how you work",
    nextTierHint: "Keep working with me and we'll hit our stride",
  },
  3: {
    label: "Hitting our stride",
    description: "We have a real working relationship",
    nextTierHint:
      "Give me more context and autonomy and we'll be fully in sync",
  },
  4: {
    label: "In sync",
    description: "Full partnership",
  },
};

export type FactCategory = "voice" | "world" | "priorities";
export type FactConfidence = "strong" | "uncertain";
export type FactSource = "onboarding" | "inferred";

export interface Fact {
  id: string;
  category: FactCategory;
  text: string;
  confidence: FactConfidence;
  source: FactSource;
}

export type CapabilityTier = "unlocked" | "next-up" | "earned";

export interface Capability {
  id: string;
  name: string;
  description: string;
  tier: CapabilityTier;
  /** Human-readable unlock requirement. */
  gate: string;
  /** Shown on "earned" tier: why, not when. */
  unlockHint?: string;
  /** Shown on "next-up" tier: e.g. "Connect Google →". */
  ctaLabel?: string;
}

export interface RelationshipState {
  version: typeof RELATIONSHIP_STATE_VERSION;
  /** Forward-compat for multi-assistant. Only one assistant is supported in v1. */
  assistantId: string;
  tier: RelationshipTier;
  /** 0-100 */
  progressPercent: number;
  facts: Fact[];
  capabilities: Capability[];
  conversationCount: number;
  /** ISO 8601 */
  hatchedDate: string;
  assistantName: string;
  userName?: string;
  /** ISO 8601 */
  updatedAt: string;
}

/**
 * Seed list of capabilities the relationship-state writer should project,
 * minus the `tier` field (which is computed at write time based on what
 * the assistant has learned and which integrations are connected).
 *
 * Order and ids are part of the public contract — the contract test
 * asserts this list matches the TDD so drift causes a failure.
 */
export const DEFAULT_CAPABILITIES: Omit<Capability, "tier">[] = [
  {
    id: "email",
    name: "Email access",
    description: "Read, draft, and manage your email",
    gate: "Connect Google",
    ctaLabel: "Connect Google →",
  },
  {
    id: "calendar",
    name: "Calendar awareness",
    description: "Know your schedule, prep for meetings",
    gate: "Connect calendar",
    ctaLabel: "Connect Calendar →",
  },
  {
    id: "slack",
    name: "Slack monitoring",
    description: "Watch channels, surface what matters",
    gate: "Set up Slack app",
    ctaLabel: "Set up Slack →",
  },
  {
    id: "voice-writing",
    name: "Write in your voice",
    description: "Draft messages and docs that sound like you",
    gate: "Usage — needs conversation history",
    unlockHint: "I need to learn how you communicate first",
  },
  {
    id: "proactive",
    name: "Proactive suggestions",
    description: "Flag things before you ask",
    gate: "Trust — needs priority understanding",
    unlockHint: "I need to understand your priorities first",
  },
  {
    id: "autonomous",
    name: "Act on your behalf",
    description: "Send messages, file things, take action",
    gate: "Trust — deepest level",
    unlockHint: "We need a deeper working relationship first",
  },
];

File: clients/shared/Models/RelationshipState.swift
import Foundation

/// Relationship state data contract.
///
/// Wire-compatible mirror of
/// `assistant/src/home/relationship-state.ts`. The TypeScript types are
/// the source of truth — any change there must be mirrored here so a
/// JSON blob produced by one side decodes byte-for-byte on the other.
public struct RelationshipState: Codable, Sendable, Equatable {
    public let version: Int
    /// Forward-compat for multi-assistant. Only one assistant is supported in v1.
    public let assistantId: String
    /// 1-4
    public let tier: Int
    /// 0-100
    public let progressPercent: Int
    public let facts: [Fact]
    public let capabilities: [Capability]
    public let conversationCount: Int
    /// ISO 8601
    public let hatchedDate: String
    public let assistantName: String
    public let userName: String?
    /// ISO 8601
    public let updatedAt: String

    public init(
        version: Int,
        assistantId: String,
        tier: Int,
        progressPercent: Int,
        facts: [Fact],
        capabilities: [Capability],
        conversationCount: Int,
        hatchedDate: String,
        assistantName: String,
        userName: String?,
        updatedAt: String
    ) {
        self.version = version
        self.assistantId = assistantId
        self.tier = tier
        self.progressPercent = progressPercent
        self.facts = facts
        self.capabilities = capabilities
        self.conversationCount = conversationCount
        self.hatchedDate = hatchedDate
        self.assistantName = assistantName
        self.userName = userName
        self.updatedAt = updatedAt
    }
}

public struct Fact: Codable, Sendable, Equatable, Identifiable {
    public let id: String
    public let category: Category
    public let text: String
    public let confidence: Confidence
    public let source: Source

    public enum Category: String, Codable, Sendable {
        case voice
        case world
        case priorities
    }

    public enum Confidence: String, Codable, Sendable {
        case strong
        case uncertain
    }

    public enum Source: String, Codable, Sendable {
        case onboarding
        case inferred
    }

    public init(
        id: String,
        category: Category,
        text: String,
        confidence: Confidence,
        source: Source
    ) {
        self.id = id
        self.category = category
        self.text = text
        self.confidence = confidence
        self.source = source
    }
}

public struct Capability: Codable, Sendable, Equatable, Identifiable {
    public let id: String
    public let name: String
    public let description: String
    public let tier: Tier
    /// Human-readable unlock requirement.
    public let gate: String
    /// Shown on `.earned` tier: why, not when.
    public let unlockHint: String?
    /// Shown on `.nextUp` tier: e.g. "Connect Google →".
    public let ctaLabel: String?

    public enum Tier: String, Codable, Sendable {
        case unlocked
        case nextUp = "next-up"
        case earned
    }

    public init(
        id: String,
        name: String,
        description: String,
        tier: Tier,
        gate: String,
        unlockHint: String?,
        ctaLabel: String?
    ) {
        self.id = id
        self.name = name
        self.description = description
        self.tier = tier
        self.gate = gate
        self.unlockHint = unlockHint
        self.ctaLabel = ctaLabel
    }
}

public enum RelationshipTier: Int, CaseIterable, Sendable {
    case gettingToKnowYou = 1
    case findingFooting = 2
    case hittingStride = 3
    case inSync = 4

    public var label: String {
        switch self {
        case .gettingToKnowYou: return "Getting to know you"
        case .findingFooting: return "Finding my footing"
        case .hittingStride: return "Hitting our stride"
        case .inSync: return "In sync"
        }
    }

    public var descriptionText: String {
        switch self {
        case .gettingToKnowYou: return "We just met — learning the basics"
        case .findingFooting: return "Starting to understand how you work"
        case .hittingStride: return "We have a real working relationship"
        case .inSync: return "Full partnership"
        }
    }

    public var nextTierHint: String? {
        switch self {
        case .gettingToKnowYou:
            return "A few more conversations and I'll start to find my footing"
        case .findingFooting:
            return "Keep working with me and we'll hit our stride"
        case .hittingStride:
            return "Give me more context and autonomy and we'll be fully in sync"
        case .inSync:
            return nil
        }
    }
}
